### PR TITLE
Enable spilling support for partial aggregation

### PR DESCRIPTION
**Pull request by zhztheplayer**
Resolves https://github.com/facebookincubator/velox/issues/7930, also be related to https://github.com/facebookincubator/velox/issues/7511

Partial aggregation currently relies on following options to reduce memory usage:

* kMaxPartialAggregationMemory
* kMaxExtendedPartialAggregationMemory

The PR adds the following one to enable flushing on partial aggregation:

* kPartialAggregationSpillEnabled (default value: false)

When `kPartialAggregationSpillEnabled` is set to true, partial aggregation will try spilling data to disk just like final aggregation can do. At the same time, flushing will be disabled.

Also, when spilling has been triggered, code path of abandoning partial aggregation will be disabled.

### CODE REVIEW

**Comment by zhztheplayer** [comment]
@mbasmanova @xiaoxmeng Would you like to review? Thanks!

**Comment by zhztheplayer** [comment]
@mbasmanova @xiaoxmeng Hi, would you please help review? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@zhztheplayer if we need to spill in partial aggr for gluten use case, we might need a separate config for this. Thanks!

**Comment by zhztheplayer** [comment]
@xiaoxmeng

File: velox/exec/GroupingSet.cpp [hunk 1]
```diff
@@ -725,6 +725,7 @@ bool GroupingSet::getOutput(
   }
 
   if (hasSpilled()) {
+    spill();
```
Comment (zhztheplayer) [inline velox/exec/GroupingSet.cpp:728]: If we want to make spilling and flushing work together, `getOutput` may have to flush in-memory and spilled data together. In the case we should spill in-memory data out then call `getOutputWithSpill`.

**Comment by zhztheplayer** [comment]
Hi @mbasmanova @xiaoxmeng , should we continue the review? Thanks!

**Review by mbasmanova — COMMENTED** [review]
@zhztheplayer Would you update PR description to describe the latest changes. The description seems outdated.

Also, please, document new configs in https://facebookincubator.github.io/velox/configs.html

File: velox/core/PlanNode.cpp [hunk 2]
```diff
@@ -233,8 +233,16 @@ bool AggregationNode::canSpill(const QueryConfig& queryConfig) const {
   }
   // TODO: add spilling for pre-grouped aggregation later:
   // https://github.com/facebookincubator/velox/issues/3264
-  return (isFinal() || isSingle()) && preGroupedKeys().empty() &&
-      queryConfig.aggregationSpillEnabled();
+  if ((isFinal() || isSingle()) && queryConfig.aggregationSpillEnabled()) {
+    return preGroupedKeys().empty();
```
Comment (mbasmanova) [inline velox/core/PlanNode.cpp:237]: This is a bit confusing. Perhaps,

```
if (preGroupedKeys().empty()) {
   return false();
}

...
```

File: velox/functions/prestosql/aggregates/ApproxPercentileAggregate.cpp [hunk 3]
```diff
@@ -650,7 +663,8 @@ class ApproxPercentileAggregate : public exec::Aggregate {
     if constexpr (checkIntermediateInputs) {
       VELOX_USER_CHECK(rowVec);
       for (int i = kPercentiles; i <= kAccuracy; ++i) {
-        VELOX_USER_CHECK(rowVec->childAt(i)->isConstantEncoding());
+        VELOX_USER_CHECK(isConstantVector(
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/ApproxPercentileAggregate.cpp:666]: Unrelated change? Would you extract it into a separate PR, provide clear description and tests?

File: velox/exec/HashAggregation.cpp [hunk 4]
```diff
@@ -135,9 +139,13 @@ void HashAggregation::addInput(RowVectorPtr input) {
   // NOTE: we should not trigger partial output flush in case of global
   // aggregation as the final aggregator will handle it the same way as the
   // partial aggregator. Hence, we have to use more memory anyway.
+  //
+  // We currently disable flushing when spilling is enabled. It's possible
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.cpp:143]: Perhaps, update comments for the new config to explain that it disables flushing.

**Review by mbasmanova — COMMENTED** [review]
Would you check if AggregationFuzzer covers the new code paths? If not, let's extend it.

File: velox/core/PlanNode.cpp (hunk 2 above)
Comment (zhztheplayer) [reply velox/core/PlanNode.cpp:237]: Sure. But would you mean the following?

```cpp
if (!preGroupedKeys().empty()) {
  return false; 
}
```

File: velox/functions/prestosql/aggregates/ApproxPercentileAggregate.cpp (hunk 3 above)
Comment (zhztheplayer) [reply velox/functions/prestosql/aggregates/ApproxPercentileAggregate.cpp:666]: It was related since without the change AggregateFuzzer would fail. 

However I'll try reverting the change since the failure might be related to flushing and spilling was once both enabled in the PR. Probably it's no longer needed.

File: velox/exec/tests/utils/AggregationFuzzer.cpp [hunk 5]
```diff
@@ -1046,6 +1063,14 @@ velox::test::ResultOrError AggregationFuzzer::execute(
           .config(core::QueryConfig::kTestingSpillPct, "100");
     }
 
+    if (injectPartialSpill) {
+      spillDirectory = exec::test::TempDirectoryPath::create();
+      builder.spillDirectory(spillDirectory->path)
+          .config(core::QueryConfig::kSpillEnabled, "true")
+          .config(core::QueryConfig::kPartialAggregationSpillEnabled, "true")
+          .config(core::QueryConfig::kTestingSpillPct, "100");
+    }
```
Comment (zhztheplayer) [inline velox/exec/tests/utils/AggregationFuzzer.cpp:1072]: Flushing will be disabled when partial spill is on so this new dimension is needed to avoid losing coverage.

File: velox/exec/HashAggregation.cpp (hunk 4 above)
Comment (zhztheplayer) [reply velox/exec/HashAggregation.cpp:143]: done

**Comment by zhztheplayer** [comment]
@mbasmanova Thanks

File: velox/core/PlanNode.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/core/PlanNode.cpp]: Oh yes. Sorry for the typo.

**Comment by zhztheplayer** [comment]
Hi @mbasmanova I've addressed current comments. Would you please take another look? Thank you.

File: velox/docs/configs.rst [hunk 6]
```diff
@@ -215,6 +215,12 @@ Spilling
      - boolean
      - true
      - When `spill_enabled` is true, determines whether HashAggregation operator can spill to disk under memory pressure.
+   * - partial_aggregation_spill_enabled
+     - boolean
+     - false
+     - When `spill_enabled` is true, determines whether the partial phase of HashAggregation operator can spill to disk under memory pressure.
+       Flushing will be disabled so max_partial_aggregation_memory and max_extended_partial_aggregation_memory will be ignored when turning the option on.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:222]: typos: " the option on. this option."

File: velox/exec/tests/utils/TempDirectoryPath.cpp [hunk 7]
```diff
@@ -28,7 +28,7 @@ std::shared_ptr<TempDirectoryPath> TempDirectoryPath::create() {
 }
 
 TempDirectoryPath::~TempDirectoryPath() {
-  LOG(INFO) << "TempDirectoryPath:: removing all files from" << path;
+  LOG(INFO) << "TempDirectoryPath:: removing all files from " << path;
```
Comment (mbasmanova) [inline velox/exec/tests/utils/TempDirectoryPath.cpp:31]: Nice fix. Thanks.

File: velox/exec/tests/SharedArbitratorTest.cpp [hunk 8]
```diff
@@ -1184,6 +1184,99 @@ TEST_F(SharedArbitrationTest, reclaimFromDistinctAggregation) {
...
 }
 
+TEST_F(SharedArbitrationTest, reclaimFromPartialAggregation) {
+  const uint64_t maxQueryCapacity = 20L << 20;
+  std::vector<RowVectorPtr> vectors = newVectors(1024, maxQueryCapacity * 2);
+  createDuckDbTable(vectors);
+  const auto spillDirectory = exec::test::TempDirectoryPath::create();
+  core::PlanNodeId partialAggNodeId;
+  core::PlanNodeId finalAggNodeId;
+  std::shared_ptr<core::QueryCtx> queryCtx = newQueryCtx(maxQueryCapacity);
```
Comment (mbasmanova) [inline velox/exec/tests/SharedArbitratorTest.cpp:1194]: auto

File: velox/exec/tests/SharedArbitratorTest.cpp [hunk 9]
```diff
@@ -1184,6 +1184,99 @@ TEST_F(SharedArbitrationTest, reclaimFromDistinctAggregation) {
...
+  auto task =
+      AssertQueryBuilder(duckDbQueryRunner_)
+          .spillDirectory(spillDirectory->path)
+          .config(core::QueryConfig::kSpillEnabled, "true")
+          .config(core::QueryConfig::kPartialAggregationSpillEnabled, "true")
+          .config(core::QueryConfig::kAggregationSpillEnabled, "true")
+          .config(
+              core::QueryConfig::kMaxPartialAggregationMemory,
+              std::to_string(1LL << 30)) // disable flush
```
Comment (mbasmanova) [inline velox/exec/tests/SharedArbitratorTest.cpp:1203]: Why do we need this? Enabling partial spilling disables flushing, no?

File: velox/exec/tests/SharedArbitratorTest.cpp [hunk 10]
```diff
@@ -1184,6 +1184,99 @@ TEST_F(SharedArbitrationTest, reclaimFromDistinctAggregation) {
...
+          .config(
+              core::QueryConfig::kMaxExtendedPartialAggregationMemory,
+              std::to_string(1LL << 30)) // disable flush
+          .config(
+              core::QueryConfig::kAbandonPartialAggregationMinPct,
```
Comment (mbasmanova) [inline velox/exec/tests/SharedArbitratorTest.cpp:1208]: > // avoid abandoning

Why do we need this?

**Review by mbasmanova — COMMENTED** [review]
@zhztheplayer Thank you for iterating. Would you update PR description to describe changes to the Fuzzer? Would you also run Fuzzer for 1h+ to make sure there are no failures. BTW, have you seen spilling in partial agg during Fuzzer runs? How often does it happen?

@duanmeng @xiaoxmeng Please, help review this PR.

File: velox/docs/configs.rst [hunk 11]
```diff
@@ -215,6 +215,12 @@ Spilling
      - boolean
      - true
      - When `spill_enabled` is true, determines whether HashAggregation operator can spill to disk under memory pressure.
+   * - partial_aggregation_spill_enabled
+     - boolean
+     - false
+     - When `spill_enabled` is true, determines whether the partial phase of HashAggregation operator can spill to disk under memory pressure.
```
Comment (mbasmanova) [inline velox/docs/configs.rst:221]: Let's explain when it could be beneficial to enable this? E.g. much fewer reducers, slow or inefficient shuffle, anything else?

File: velox/exec/tests/SharedArbitratorTest.cpp (hunk 10 above)
Comment (zhztheplayer) [reply velox/exec/tests/SharedArbitratorTest.cpp:1208]: The test cases meant to test against spilling but if partial aggregation got abandoned then spill will hardly happen.

File: velox/docs/configs.rst (hunk 11 above)
Comment (zhztheplayer) [reply velox/docs/configs.rst:221]: Can we just describe the result when turning the option on? For example, just note user the size of data emitted from partial aggregation would be reduced as much as possible if setting to `true`. Users will know it should be on if they rely on inefficient shuffle or have other specific reasons.

File: velox/exec/tests/SharedArbitratorTest.cpp (hunk 9 above)
Comment (zhztheplayer) [reply velox/exec/tests/SharedArbitratorTest.cpp:1203]: The test case explicitly eliminates the impact of flushing by disabling it, despites whether we had made flushing and spilling exclusive. In future we may decide to make flushing and spilling work together but this case should still work and still test against flushing independently in that case.

The next test case will ensure flushing is disabled by spilling. Would this make sense to you?

File: velox/exec/HashAggregation.cpp [hunk 12]
```diff
@@ -113,6 +113,10 @@ void HashAggregation::initialize() {
 
 bool HashAggregation::abandonPartialAggregationEarly(int64_t numOutput) const {
   VELOX_CHECK(isPartialOutput_ && !isGlobal_);
+  if (groupingSet_->hasSpilled()) {
+    // Once spilling kicked in, disable the abandoning code path.
```
Comment (duanmeng) [inline velox/exec/HashAggregation.cpp:117]: Does this mean to simplify the getOutput processing? Could you please help to add a comment on why disable the abandoning code path?

**Review by duanmeng — COMMENTED** [review]
@zhztheplayer This PR is crucial to batch / ETL queries, 👍 . Perhaps we could add some test cases in the `AggregationTest`.

**Comment by duanmeng** [comment]
@mbasmanova @xiaoxmeng Do we need two separate spill-enable flags if our goal is to support spilling for aggregation in all equivalent query plans? WDYT?

File: velox/exec/HashAggregation.cpp (hunk 12 above)
Comment (zhztheplayer) [reply velox/exec/HashAggregation.cpp:117]: Spilling would invalidate the condition check for abandoning. Say spilled rows will not count to `numOutput` based on current way of computation. I'll add some comments.

**Comment by zhztheplayer** [comment]
> @zhztheplayer This PR is crucial to batch / ETL queries, 👍 . Perhaps we could add some test cases in the `AggregationTest`.

Thanks @duanmeng . I've added some tests to `AggregationTest`.

**Comment by zhztheplayer** [comment]
> @mbasmanova @xiaoxmeng Do we need two separate spill-enable flags if our goal is to support spilling for aggregation in all equivalent query plans? WDYT?

Let's keep them separate for the initial version? Spilling for partial aggregation should be turned off by default since it disables flushing. While for regular aggregation spilling is on by default.

**Comment by zhztheplayer** [comment]
@mbasmanova @xiaoxmeng @duanmeng 

After going through some history about partial aggregation and companion functions (especially https://github.com/facebookincubator/velox/issues/4412 and https://github.com/facebookincubator/velox/pull/4566), I am feeling that the topic is on the wrong direction. So I am about to close this.

Reason:

Spark's partial aggregation actually doesn't support flushing so ideally in Gluten we should not use Velox (probably Presto also)'s `PartialAggregation`. Velox's `FinalAggregation + Companion functions` is better for Gluten to choose. Since final aggregation already supports spilling so no extra effort is needed in my opinion. 

`PartialAggregation` in Velox as a concept should consist of optimizations when a regular aggregation doesn't have to emit full-aggregated data. Spark doesn't have a plan type mapping to this functionality yet.

**Comment by zhztheplayer** [comment]
Closing. Please reopen if anyone thinks it's still needed.

**Comment by mbasmanova** [comment]
@zhztheplayer Thank you for the update. It comes as a bit of surprise, but I guess it makes sense. Would you also comment on the 2 related issues and close these?

> Resolves https://github.com/facebookincubator/velox/issues/7930, also be related to https://github.com/facebookincubator/velox/issues/7511

**Comment by mbasmanova** [comment]
@zhztheplayer Curious, what prompted you to take another look and go "through some history about partial aggregation and companion functions".

**Comment by zhztheplayer** [comment]
> @zhztheplayer Curious, what prompted you to take another look and go "through some history about partial aggregation and companion functions".

One of Gluten's users reported an issue that expression `count(distinct l_orderedkey)` returns wrong result in Gluten. Spark generated the following plan:

```
+- HashAggregate(keys=[], functions=[count(distinct l_orderkey#114L)])
   +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=140]
      +- (Partial) HashAggregate(keys=[], functions=[partial_count(distinct l_orderkey#114L)])    // <- node 2
         +- (Partial) HashAggregate(keys=[l_orderkey#114L], functions=[])   // <- node 1
            +- Exchange hashpartitioning(l_orderkey#114L, 100), ENSURE_REQUIREMENTS, [plan_id=136]
               +- HashAggregate(keys=[l_orderkey#114L], functions=[])
                  +- FileScan
```

The plan node I marked as `node 2` is using `node 1`'s  output as distinct input to do `partial_count`. However since `node 1` is now mapped to Velox's `PartialAggregation`, it may flush intermediate data at any time. That makes `node 2` output wrong count result.

Looks to be that the user's issue is related to incorrect mapping from Spark's `Aggregation + partial function` to Velox's `PartialAggregation`. If we correct the mapping as from Spark's `Aggregation + partial function` to Velox's `FinalAggregation + partial companion function`, then we can solve this count+distinct issue and don't have to add spilling support to partial aggregation either. I'm still working with @rui-mo to try to do some refactors to Gluten to use the new mapping, but anyway I think this topic can be closed since it was driven by issue related to Gluten's non-optimal design.
